src/EntityManagerFactory.php

```diff
@@ -318,14 +325,32 @@ protected function setCustomFunctions(Configuration $configuration)
      */
     protected function setCacheSettings(Configuration $configuration)
     {
-        if ($namespace = $this->config->get('doctrine.cache.namespace', null)) {
-            $this->cache->driver()->setNamespace($namespace);
-        }
+        $configuration->setQueryCacheImpl($this->applyNamedCacheConfiguration('query'));
+        $configuration->setResultCacheImpl($this->applyNamedCacheConfiguration('result'));
+        $configuration->setMetadataCacheImpl($this->applyNamedCacheConfiguration('metadata'));
 
         $this->setSecondLevelCaching($configuration);
     }
 
     /**
+     * @param  string $cacheName
+     * @return mixed
+     */
+    private function applyNamedCacheConfiguration($cacheName)
+    {
+        $defaultDriver    = $this->config->get('doctrine.cache.default', $this->defaultCache['type']);
+        $defaultNamespace = $this->config->get('doctrine.cache.namespace', $this->defaultCache['namespace']);
+
+        $driver = $this->config->get('doctrine.cache.' . $cacheName . '.driver', $defaultDriver);
```

Comment:
$cache = $this->cache->driver($driver);

if ($namespace = $this->config->get('doctrine.cache.' . $cacheName . '.namespace', $defaultNamespace)) {
    $cache->setNamespace($namespace);
}

return $cache;